sdb/parser.py

```diff
@@ -0,0 +1,206 @@
+#
+# Copyright 2020 Delphix
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""
+This module contains the logic for the tokenization and parsing
```

Comment:
Generally speaking I am somewhat leery of using a hand-rolled parser instead of a generated one; I assume you looked at options for parser generators and found them lacking in some way? It might be good to document that somewhere.

Reply:
I decided to write the following block comment below the module docstring:
```
#
# Why Roll Our Own Parser?
#
# Our grammar in its current state could be implemented with shlex() that is
# part of the standard library if we applied some workarounds to it. That said
# the code wouldn't be clean, it would be hard to add new rules (workarounds
# on top of workaroudns) and providing helpful error messages would be hard.
#
# In terms of external parsing libraries, the following ones were considered:
#   * PLY (Python Lex-Yacc)
#   * SLY (Sly Lex-Yacc)
#   * Lark
#
# PLY attempts to model traditional Lex & Yacc and it does come with a lot of
# their baggage. There is a lot of global state, that we'd either need to
# recreate (e.g. regenerate the grammar) every time an SDB command is issued,
# or alternatively we'd need to keep track of a few global objects and reset
# their metadata in both success and error code paths. The latter is not that
# bad but it can be very invasive in parts of the code base where we really
# shouldn't care about parsing. In addition, error-handling isn't great and
# there is a lot of boilerplate and magic to it.
#
# SLY is an improved version of PLY that deals with most issues of global
# state and boilerplace code. The error-handling is still not optimal but a
# lot better, optimizing for common cases. SLY would provide a reasonable
# alternative implementation to our hand-written parser but it wasn't chosen
# mainly for one reason. It tries to optimize for traditional full-fledged
# languages which results in a few workarounds given SDB's simplistic but
# quirky command language.
#
# Lark is probably the best option compared to the above in terms of features,
# ergonomics like error-handling, and clean parser code. The only drawback of
# this library in the context of SDB is that it is hard to debug incorrect
# grammars - the grammar is generally one whole string and if it is wrong the
# resuting stack traces end up showing methods in the library, not in the
# code that the consumer of the library wrote (which is what would geenrally
# happen with SLY). This is not a big deal in general but for SDB we still
# haven't finalized all the command language features (i.e. subshells or
# defining alias commands in the runtime) and our grammar isn't stable yet.
#
# Our hand-written parser below has a small implementation (less than 100
# lines of code without the comments), provides friendly error messages,
# and it falls cleanly to our existing code. As SDB's command language
# grows and gets more stable it should be easy to replace the existing
# parser with a library like Lark.
#
```
Is this close to what you were looking for? If not let me know and we can change it.

Reply:
Looks great to me! Thanks for giving such a thorough breakdown.